src/quicklist.c

```diff
@@ -1295,15 +1443,25 @@ int quicklistIndex(const quicklist *quicklist, const long long idx,
 void quicklistRotate(quicklist *quicklist) {
     if (quicklist->count <= 1)
         return;
-
     /* First, get the tail entry */
-    unsigned char *p = ziplistIndex(quicklist->tail->zl, -1);
+    unsigned char *p = NULL;
     unsigned char *value, *tmp;
     long long longval;
-    unsigned int sz;
+    size_t sz;
     char longstr[32] = {0};
-    ziplistGet(p, &tmp, &sz, &longval);
 
+    if(unlikely(quicklist->tail->container == QUICKLIST_NODE_CONTAINER_NONE))
+    {
+        /* Add tail entry to head (must happen before tail is deleted). */
+        quicklistPushHead(quicklist, quicklist->tail->entry, quicklist->tail->sz);
+        /* Remove tail entry. */
+        quicklistDelIndex(quicklist, quicklist->tail, &p);
+        return;
+    }
+
+    p = ziplistIndex(quicklist->tail->zl, -1);
+
+    ziplistGet(p, &tmp, (unsigned int*) &sz, &longval);
```

Comment:
i just wanna confirm..
we wanna remove one entry from the tail and put it at the head.
the case you handled above is where the tail is "plain", so you just need to move a full node.

but there's another case, where the tail is a ziplist, and you wanna move an entry to the head, but the head is "plain" (not a ziplist). is that case ok because quicklistPushHead handles that?

Reply:
yes it is handled by quicklistPushHead

Reply:
let's make sure there's a unit-test that covers it.